futures-util/src/io/write_all_vectored.rs

```diff
@@ -190,4 +190,25 @@ mod tests {
             assert_eq!(&*dst.written, &*wanted);
         }
     }
+
+    #[test]
+    fn test_write_all_vectored_distinct_lifetimes() {
+        let waker = noop_waker();
+        let mut cx = Context::from_waker(&waker);
+        let mut dst = test_writer(2, 2);
+
+        {
+            // Force the lifetimes of the underlying data and IOSlice to be unequal
+            let data = vec![1, 2, 3, 4];
+            {
+                let mut slices: Vec<_> = data.chunks(2).map(IoSlice::new).collect();
+                let mut future = dst.write_all_vectored(&mut *slices);
```

Comment:
Currently, this test seems to work fine without this patch.


```suggestion
        struct WrapVec<T>(Vec<T>);
        impl<T> Drop for WrapVec<T> {
            fn drop(&mut self) {}
        }

        let waker = noop_waker();
        let mut cx = Context::from_waker(&waker);
        let mut dst = test_writer(2, 2);

        {
            // Force the lifetimes of the underlying data and IOSlice to be unequal
            let data = vec![1, 2, 3, 4];
            {
                let mut slices = WrapVec(data.chunks(2).map(IoSlice::new).collect());
                let mut future = dst.write_all_vectored(&mut *slices.0);
```

Reply:
You are quite right it does, oops. I don't quite follow why your revision correctly fails in the original version where mine works, but I've accepted it just now anyway